src/helpers.php

```diff
@@ -246,3 +246,27 @@ function is_countable($obj)
         return is_array($obj) || $obj instanceof Countable;
     }
 }
+
+if (! function_exists('old_input_value')) {
+    /**
+     * This function allows us to setup a default value in case there is an old value, but is forcelly null
+     * by Laravel middleware ConvertEmptyStringsToNull.
+     *
+     * @param  string  $field_name
+     * @param  mixed  $fallback
+     * @return mixed
+     */
+    function old_input_value($input_name, $fallback_value = null)
+    {
+        $input_name = square_brackets_to_dots($input_name);
+        $old_inputs = session()->getOldInput();
+
+        // if the input name is present in the old inputs we need to return earlier and not in a coalescing chain
+        // otherwise `null` aka empty will not pass the condition and the field value would be returned.
+        if (\Arr::has($old_inputs, $input_name)) {
+            return \Arr::get($old_inputs, $input_name);
+        }
+
+        return $fallback_value;
+    }
+}
```

Comment:
I see there are a few fields that are NOT using `old_input_value()` now. 
I'm not saying they _should_... I'm asking... why aren't they - is this intentional or a mistake?
- [ ] browse_multiple
- [ ] checklist_dependency
- [ ] enum
- [x] password - it never pre-populates the value so I understand why this one doesn't; 
- [ ] select2_from_array
- [ ] select2_multiple
- [ ] select_from_array
- [ ] select_multiple
- [ ] upload_multiple

Do they still have "_the bug_"? That's what matters after all. If they don't... we're good, no need to refactor to use `old_input_value()`. If they _do_ have the bug, we should fix that, and the cleanest would probably be using `old_input_value()`.

Reply:
@tabacitu I am taking care of this ... and I think I found why explicitly pass the fallback (what field should show when empty) to the `old_input_value` function. It's what should be returned whe input is present in the `old()` but it's null. 

If we chain `value ?? default ?? fallback` as the "fallback", if we are editing an entry we will never reach the `default` or `fallback` because the `value` would already exists. And when you submit the input empty it will show the `value` instead of the fallback as it should. 

A written example:
- text field in monster crud. 
- create a monster and then go edit it.
- clear the text field so the validation fails. 
- Now you will get whatever is present in the `old()` - FINE, IT WILL BE EMPTY
- But it will fail for example in fields that require the "fallback" to be `[]` or that don't parse the "null" well. 

We can: 
- Use a third parameter as a solo fallback. `old_input_value(field_name, default OR value, fallback)`
```php
function old_input_value($input_name, $valueOrDefault, $fallback)
    {
        $input_name = square_brackets_to_dots($input_name);
        $old_inputs = session()->getOldInput();

        // if the input name is present in the old inputs we need to return earlier and not in a coalescing chain
        // otherwise `null` aka empty will not pass the condition and the field value would be returned.
        if (\Arr::has($old_inputs, $input_name)) {
            return \Arr::get($old_inputs, $input_name) ?? $fallback; //so if it's null here, we want to return the field "empty state"
        }

        return $valueOrDefault;
    }
```

Some parts of my body were telling me something isn't rigth when I was refactoring ... We had talked about this issue sometimes, and as I recall we started doing as you suggested the refactor (how it is currently implemented) and then ended-up sending more stuff to the function.

Let me know what you think before I refactor this. 

Pedro

Reply:
```php

has_old_input('title') ? get_old_input('title', $fallback) : $value ?? $default ?? $fallback;

old_input_fallback_or_null('title', '') ?? $value ?? $default ?? '';
old_input_value('title', '[]') ?? $value ?? $default ?? '[]';
old_empty_or_fallback('title', [], $value ?? $default);
```

Reply:
@pxpm let's do it with `old_empty_or_fallback()`. The name isn't very clear... but it's clear ENOUGH. And if to the dev it isn't clear enough, they can click on it to go to definition, where we'll provide a comprehensive explanation. So:

```php

/**
 * This method is an alternative to Laravel's old() helper, which mistakenly
 * returns NULL it two cases:
 * - if there is an old value, and it was empty or null
 * - if there is no old value
 *
 * In contrast, this method will return:
 * - the old value, if there is one for that input;
 * - the second parameter, if there is no old value for that input;
 * - the third parameter, if there is no old input at all;
 */
function old_empty_or_fallback($input_name, $empty, $fallback)
    {
        $input_name = square_brackets_to_dots($input_name);
        $old_inputs = session()->getOldInput();

        // if the input name is present in the old inputs we need to return earlier and not in a coalescing chain
        // otherwise `null` aka empty will not pass the condition and the field value would be returned.
        if (\Arr::has($old_inputs, $input_name)) {
            return \Arr::get($old_inputs, $input_name) ?? $empty; // so if it's null here, we want to return the field "empty state"
        }

        return $fallback;
    }
```

I think this is the best we can do for now. Strikes an "ok" balance between simplicity, legibility and smaller breaking change. Don't you think? 💪

Reply:
Pedro, I noticed you refactored and encountered a big problem: we can't call `old_empty_or_fallback('title', [], $value ?? $default);` because there IS NO DEFAULT most of the time so it'll error. So we end up calling `old_empty_or_fallback('title', [], $value ?? $default ?? '');` anyway. That means we end up typing the empty string twice anyway...

If that's true, **change of plans**. Let's do this instead:

```php
old_empty_or_null('title', '') ?? $value ?? $default ?? '';
```

It's ugly... but it works (at least from my tests). And it's a one-liner. Here's the function I used, with a very good description (I think):

```php
if (! function_exists('old_empty_or_null')) {
    /**
     * This method is an alternative to Laravel's old() helper, which mistakenly
     * returns NULL it two cases:
     * - if there is an old value, and it was empty or null
     * - if there is no old value
     * (this is because of the ConvertsEmptyStringsToNull middleware)
     *
     * In contrast, this method will return:
     * - the old value, if there actually is an old value for that key;
     * - the second parameter, if there is no old value for that key, but it was empty string or null;
     * - null, if there is no old value at all for that key;
     *
     * @param  string $key
     * @param  array|string $empty_value
     * @return mixed
     */
    function old_empty_or_null($key, $empty_value = '')
    {
        $key = square_brackets_to_dots($key);
        $old_inputs = session()->getOldInput();

        // if the input name is present in the old inputs we need to return earlier and not in a coalescing chain
        // otherwise `null` aka empty will not pass the condition and the field value would be returned.
        if (\Arr::has($old_inputs, $key)) {
            return \Arr::get($old_inputs, $key) ?? $empty_value;
        }

        return null;
    }
}
```

This will solve all our problems, right?

Reply:
Indeed @tabacitu 

I stopped yesterday but it was late and I was tired so I let it boil down during the nigth to come with fresh eye today.

Totally agree with your reasoning. 

Back to you here in a few 👍